EnsureRemoveAll, RecursiveUnmount: don't call Mounted around Unmount

kolyshkin opened the pull request:
### "pkg/system".EnsureRemoveAll()

* Call to `mount.Mounted()` is very expensive and it's redundant
   to call it before `Unmount()`.

* Calling `mount.Mounted()` after an error from `Unmount()` is
   questionable -- if unmounting failed, the mount is probably
   still there anyway, it doesn't make sense to check it.

### "pkg/system".RecursiveUnmount()

* same as item 2 above

This should result in faster code with no change in functionality.

@cpuguy83 PTAL

kolyshkin committed:
pkg/system.EnsureRemoveAll: don't call Mounted

1. Call to mount.Mounted() is very expensive and it's redundant
   to call it before Unmount().

2. Calling mount.Mounted() after an error from Unmount() is
   questionable -- if umount failed, the mount is probably
   still there anyway, it doesn't make sense to check it.

This should result in faster code with no change in functionality.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

cpuguy83 commented on pkg/system/rm.go:
```diff
@@ -63,12 +63,8 @@ func EnsureRemoveAll(dir string) error {
-				}
-			}
+		if e := mount.Unmount(pe.Path); e != nil {
+			return errors.Wrapf(e, "error while removing %s", dir)
```
Just a thought, check for EINVAL here and continue in that case.

kolyshkin replied:
`mount.Unmount()` already filters out `EINVAL`

kolyshkin committed:
pkg/mount.RecursiveUnmount(): don't call Mounted()

Calling mount.Mounted() after an error from Unmount() is
questionable -- if umount failed, the mount is probably
still there anyway, it doesn't make sense to check it.

Signed-off-by: Kir Kolyshkin <kolyshkin@gmail.com>

kolyshkin commented:
Added a second commit doing the same in `RecursiveUnmount()`.

thaJeztah commented on pkg/mount/mount.go:
```diff
@@ -144,9 +144,7 @@ func RecursiveUnmount(target string) error {
-				if mounted, e := Mounted(m.Mountpoint); e != nil || mounted {
-					return err
-				}
+				return err
```
should we wrap this one as well, or does the produced error message contain enough information?

```suggestion
				return errors.Wrapf(err, "failed to unmount submount %s", m.Mountpoint)
```

cpuguy83 replied:
It does give details about the error.

kolyshkin replied:
> should we wrap this one

@thaJeztah fortunately, since #38068 we don't have to :man_dancing:

thaJeztah replied:
Ah, cool

cpuguy83 reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM

kolyshkin commented:
CI failure is unrelated, restarting

> [2020-03-06T23:23:36.283Z] fatal: unable to access 'https://github.com/docker/buildx.git/': Could not resolve host: github.com

thaJeztah commented:
@kolyshkin can you try rebasing? I tried restarting a few times, but that didn't work; recent PR's went green though, so perhaps there was a fix in master

kolyshkin commented:
rebased on top of the current master to potentially help with CI

thaJeztah commented:
Ignoring s390x, as it's having issues
